man/systemd-nspawn.xml

```diff
@@ -399,6 +399,18 @@
         is read from it and automatically used, also as formatted hexadecimal characters.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--root-hash-sig=</option></term>
+
+        <listitem><para>Takes a path to a PKCS7 formatted signature of the <option>--root-hash=</option> option.
+        If libcryptsetup (minimum version 2.3.0) and the kernel (minimum version 5.4 with CONFIG_DM_VERITY_VERIFY_ROOTHASH_SIG
+        enabled) support it, the dm-verity volume will only be opened if the signature of the root hash hex string is valid
+        and done by a public key present in the kernel keyring. If this option is not specified, but a file with the
+        <filename>.p7</filename> suffix is found next to the image file, bearing otherwise the same name (except if the image
+        has the <filename>.raw</filename> suffix, in which case the signature file must not have it in its name), the
+        signature is read from it and automatically used.</para></listitem>
+      </varlistentry>
```

Comment:
Hmm, so this pkcs7 signature, that's just the signature and the root hash needs to be specified seperately?

How does a pkcs7 file look like? Can't we optionally allow this to be specified literally? i.e. why require this to be placed in an external file if we can do literal stuff too, like we can do for the root hash itself? I mean, in an Ed25519 world signatures can be really short...

Reply:
Yes, it's just the signature of the root hash (this is driven by the kernel interface). I am not a super openssl guru so please bear with me - the kernel expects it in binary format (commonly referred to as DER with .cer extension, which I think is a base64 binary encoding of some sort), and this is how a sig that the kernel likes is created:

```
openssl smime -sign -in img.roothash -nocerts -inkey /path/to/key -signer /path/to/crt -noattr -binary -outform der -out img.p7
```

Decoding that in hex doesn't look too bad, so perhaps supporting it literally would be doable:

```
$ xxd -p img.squashfs.p7
308201b506092a864886f70d010702a08201a6308201a2020101310f300d
06096086480165030402010500300b06092a864886f70d0107013182017d
3082017902010130523045310b3009060355040613024155311330110603
5504080c0a536f6d652d53746174653121301f060355040a0c18496e7465
726e6574205769646769747320507479204c7464020900982aff96a226db
44300d06096086480165030402010500300d06092a864886f70d01010105
000482010078dbaf110c806e72be81f42d3844f499cb4e958d89355edce3
acd08a53de5d8f1ab62d11f4ebc90ba84f0d85d6eeb9c77abccec4799094
1645a1383867940ee0934c6b9096b38b2af931fee25c24906adefaf8b658
c8598afd2cf143e674b00291ad71898e654c70f696aaf476b0f0f495c1e0
a197c2746e665fa788c0cffdbb80dda82337c2eb8beb340ca52f643b317f
cdf06f62222c66358c89f533329b9def2ce42cdeb1d34f375d0b85347f61
fab36450cc091f2f7521218729727c4ef2fa0f38f345a889158a171b4469
5cbbda00dcf00bb05161f709e58755a52f1ef0665ead039f0a352187c565
9a8677cc3f0eaebcbdf83197af4e9bd5441d3de8d7
```

Reply:
One can also choose to support plain text/PEM format, but then you'd need to use openssl-y machinery to detect the format and convert it on the fly. Not sure if it's worth it? If required I can look into it

Reply:
Usually we encode "file-like" stuff in base64 if we want to embedd it in settings, see StandardInputData= for example. I think that should be fine here, no? It's trivial to parse for us since we have unbase64mem(), hence should be natural to have from day one?

Reply:
Sure thing. I was already testing with accepting hex encoding, will change to base64 instead. Do you want base64 for command line parameters as well (nspawn/systemd-dissect) ?

Reply:
yes, please let's keep this as much in sync in behaviour as we can. If it's base64 for this in one place, it should also be bas64 at all other places where we talk about the same concept.

Reply:
Just realized: given ```/``` is a valid character in base64, it could happen that a signature starts with it, and thus we mistake it for a path.
What about base32 instead?

Reply:
maybe require base64 data to be prefixed with `base64:` or so. There is prior art like this, i.e. if you encode a binary key in /etc/crypttab for example...

Reply:
Ok, sounds good - all done and pushed